content/learning-paths/mobile-graphics-and-gaming/android_halide/fusion.md

```diff
@@ -0,0 +1,349 @@
+
+The presented code demonstrates operation fusion, combining computational stages via compute_at. Halide also offers a separate optimization technique called loop fusion (via the fuse scheduling directive), which specifically merges two loop indices into a single loop variable. While the use of compute_at sufficiently addresses memory and computational efficiency concerns, explicit loop fusion can be used for additional optimization, especially to enhance cache locality or simplify parallelization.
+
+Below is a snippet explicitly demonstrating how the fuse directive can be applied in the context of the existing pipeline:
+
+```cpp
+// Existing definitions
+Halide::Var x("x"), y("y");
+
+// Gaussian blur definition remains as before.
+Halide::Func blur("blur");
+Halide::RDom r(0, 3, 0, 3);
+Halide::Expr weight = Halide::select(
+    (r.x == 1 && r.y == 1), 4,
+    (r.x == 1 || r.y == 1), 2,
+    1
+);
+
+Halide::Expr offsetX = clampCoord(x + (r.x - 1), width);
+Halide::Expr offsetY = clampCoord(y + (r.y - 1), height);
+Halide::Expr val = Halide::cast<int32_t>(input(offsetX, offsetY)) * weight;
+blur(x, y) = Halide::cast<uint8_t>(Halide::sum(val) / 16);
+
+// Thresholding definition
+Halide::Func thresholded("thresholded");
+thresholded(x, y) = Halide::select(blur(x, y) > 128, 255, 0);
+
+// ---- Explicit use of loop fusion (fuse) ----
+Halide::Var xy("xy");  // New fused loop variable.
+
+// Fuse x and y dimensions into one linearized dimension (xy).
+thresholded.fuse(x, y, xy);
+
+// Parallelize the fused loop for improved performance.
+thresholded.parallel(xy);
+
+// Compute blur within thresholded, ensuring operation fusion.
+blur.compute_at(thresholded, xy);
+```
+
+This explicit loop-fusion example demonstrates additional optimizations achievable with Halide, clearly differentiating between operation fusion (compute_at) and loop fusion (fuse).
+
+## When to use operation fusion
+Operation fusion is especially beneficial for pipelines involving multiple sequential element-wise operations. These operations perform independent transformations on individual pixels without requiring neighboring data. Element-wise operations benefit greatly from fusion since they avoid the overhead associated with repeatedly storing and loading intermediate results, significantly reducing memory bandwidth usage.
+
+Ideal use-cases for fusion are 
+* Pixel intensity normalization (scaling and shifting)
+* Color-space transformations (e.g., RGB to grayscale conversion)
+* Simple arithmetic or logical operations applied pixel-by-pixel
+
+Here’s an illustrative example:
+```cpp
+Halide::Var x("x"), y("y");
+Halide::Func scale("scale"), offset("offset"), clamp_result("clamp_result");
+
+// Element-wise transformations
+scale(x, y) = input(x, y) * 1.5f;
+offset(x, y) = scale(x, y) + 10.0f;
+clamp_result(x, y) = Halide::clamp(offset(x, y), 0.0f, 255.0f);
+
+// Apply fusion for maximum efficiency
+scale.compute_at(clamp_result, x);
```

Comment:
I don't think we use `compute_at` to schedule fusion. Instead, we usually just leave it to default Halide scheduling (which is compute_inline() = fusion)